Users with view but not write don't get notification that they cnnot create annotations

* Search [mailing list](http://gmod.827538.n3.nabble.com/Apollo-f815553.html) or email list ```apollo@lists.lbl.gov``` if a general setup question. 

* Provide what you were doing and what you expected to see.  Screenshots, directory, config files are a plus if relevant.
 ```text
I am testing user permission use cases.  This test was for a user with view but not write.  
As expected, the user could see but not create annotations. The interface allows click to highlight 
and right click to create (and the option to drag to the User-Created Annotations panel).  Since the 
options are visible and not greyed out, the impression is that the user should be able to perform 
the action.  

My expectation is that if a user has view but not write, they should be able to see the User-Created 
Annotations track, but options to create annotations, either through clicking and dragging or right 
clicking should not be available. 
```
* Provide the javascript console log output generated from the action. 
```text
 No errors present in console or in the Apollo interface
```
* Provide the server log output generated from the action (typically ```catalina.out```).
 ```shell
2017-06-13 11:37:35,022 [clientInboundChannel-3] ERROR springwebsocket.GrailsSimpAnnotationMethodMessageHandler  - Unhandled exception
java.lang.reflect.InvocationTargetException
	at org.bbop.apollo.AnnotationEditorController$__do_annotationEditor_closure4$_closure5$_closure6.doCall(AnnotationEditorController.groovy:1145)
	at org.bbop.apollo.AnnotationEditorController$__do_annotationEditor_closure4$_closure5.doCall(AnnotationEditorController.groovy:1144)
	at org.bbop.apollo.AnnotationEditorController$__do_annotationEditor_closure4.doCall(AnnotationEditorController.groovy:1140)
	at groovyx.gpars.group.PGroup$3.call(PGroup.java:289)
	at groovyx.gpars.group.PGroup$4.run(PGroup.java:313)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.bbop.apollo.AnnotationException: You have insufficient permissions [read < write] to perform this operation
	at org.bbop.apollo.PermissionService.$tt__checkPermissions(PermissionService.groovy:340)
	at org.bbop.apollo.RequestHandlingService.$tt__addTranscript(RequestHandlingService.groovy:624)
	... 8 more
```

--- a/client/apollo/js/View/Track/AnnotTrack.js
+++ b/client/apollo/js/View/Track/AnnotTrack.js
@@ -416,7 +416,7 @@ define([
             isJSON: function(str) {
                 if(!str) return false ;
                 try{
-                    j = JSON.parse(JSON.parse(str));
+                    JSON.parse(JSON.parse(str));
                     return true
                 }
                 catch(e){

--- a/client/apollo/js/View/Track/AnnotTrack.js
+++ b/client/apollo/js/View/Track/AnnotTrack.js
@@ -432,8 +432,10 @@ define([
                         changeData = JSON.parse(JSON.parse(message.body));
                     }
                     else{
-                        // we assume its an error message
-                        alert('Error: '+ message.body);
+                        changeData = {} ;
+                        changeData.operation = 'ERROR';
+                        changeData.username = track.username ;
+                        changeData.error_message = message.body ;
                     }
 
                     if (track.verbose_server_notification) {

--- a/client/apollo/js/View/Track/AnnotTrack.js
+++ b/client/apollo/js/View/Track/AnnotTrack.js
@@ -412,12 +412,29 @@ define([
                 });
                 console.log('connection established');
             },
+            // TODO: need a better function and to not have it be double-encoded
+            isJSON: function(str) {
+                if(!str) return false ;
+                try{
+                    j = JSON.parse(JSON.parse(str));
+                    return true
+                }
+                catch(e){
+                    return false
+                }
+            },
             annotationNotification: function (message) {
                 var track = this;
                 var changeData;
 
                 try {
-                    changeData = JSON.parse(JSON.parse(message.body));
+                    if(this.isJSON(message.body)){
+                        changeData = JSON.parse(JSON.parse(message.body));
+                    }
+                    else{
+                        // we assume its an error message
+                        alert('Error: '+ message.body);
+                    }
 
                     if (track.verbose_server_notification) {
                         console.log(changeData.operation + " command from server: ");

--- a/grails-app/controllers/org/bbop/apollo/AnnotationEditorController.groovy
+++ b/grails-app/controllers/org/bbop/apollo/AnnotationEditorController.groovy
@@ -48,6 +48,7 @@ class AnnotationEditorController extends AbstractApolloController implements Ann
     def preferenceService
     def sequenceSearchService
     def featureEventService
+    def brokerMessagingTemplate
 
 
     def index() {
@@ -1162,6 +1163,8 @@ class AnnotationEditorController extends AbstractApolloController implements Ann
         } catch (AnnotationException ae) {
             // TODO: should be returning nothing, but then broadcasting specifically to this user
             log.error("Error for user ${principal?.name} when exexecting ${inputString}"+ ae?.message)
+//            brokerMessagingTemplate.convertAndSend "/topic/AnnotationNotification/" + sequence.organismId + "/" + sequence.id, returnString
+            brokerMessagingTemplate.convertAndSend "/topic/AnnotationNotification/user/" + principal?.name, ae?.message
             return sendError(ae, principal.name)
         }
 

--- a/grails-app/services/org/bbop/apollo/PermissionService.groovy
+++ b/grails-app/services/org/bbop/apollo/PermissionService.groovy
@@ -449,6 +449,10 @@ class PermissionService {
             log.error("User ${jsonObject.username} does not exist in the database.")
             return false
         }
+        if (jsonObject.error_message) {
+            log.error("Error with user permissions ${user.username}:  ${jsonObject.error_message}")
+            return false
+        }
 
         // if the rank required is less than administrator than ask if they are an administrator
         if (PermissionEnum.ADMINISTRATE.rank < permissionEnum.rank ) {

--- a/grails-app/services/org/bbop/apollo/RequestHandlingService.groovy
+++ b/grails-app/services/org/bbop/apollo/RequestHandlingService.groovy
@@ -35,7 +35,6 @@ class RequestHandlingService {
     def nonCanonicalSplitSiteService
     def configWrapperService
     def nameService
-    def overlapperService
     def permissionService
     def preferenceService
     def featurePropertyService

